Completed implementation

jordan-goo opened the pull request:
## Changes
Adds FilterBuilderView, styling, and example usage.

## Testing / Demo:

https://github.com/jordan-goo/jordan-saved-filter-ui/assets/10941066/c06fcee3-ea7c-496e-acc3-62e93c64c0b5

## Decisions / Talking points

- I focused on behavior and UX for the most part, I really wanted to get the intuitive controls and keyboard navigation working well.
- Didn't worry too much about styling, usually an existing component library is used anyways so I used a single css file instead of pulling in something more complex like Tailwind or a css-in-js library.
- I decided to keep the whole thing dependency free, while there are probably tons of "Autocomplete" components or "Input" components available on npm I wanted to make sure what I had worked well together and I didn't want to over complicate things.
- Chose to use native html controls wherever possible and not use state-controlled inputs (except in the number input), I think this keeps internal state a lot cleaner and takes advantage of a lot of browser features that already exist.
- Most of the internal state is pretty simple, one exception is the use of `useReducer` in `FilterBuilderView`. I think I could have used a normal `useState` hook, but knowing I would want to expand it in the future to add "sort" and "group by", I think the reducer makes it a lot easier to extend.

jordan-goo committed:
Completed implementation

Adds FilterBuilderView, styling, and example usage.

Update README with local instructions

Reorganize files into directories

Separate into dedicated components folders

jfmaggie commented:
@jordan-goo i did a bit testing https://www.loom.com/share/9be614fdddc94ba8b669e6d0cd899b32 do you mind addressing the couple questions in the video?

jordan-goo committed:
Fix bug when editing date input to be invalid date

Instead of assuming all dates are valid, we now check for valid dates before triggering onChange function in the DateInput component. I also added an innerInputRef to the DateInput component so we can reset it to the last valid input after losing focus.

jordan-goo commented:
Thanks for the feedback @jfmaggie and good catch on the bug! I pushed another commit which fixes the issue. It will now only trigger the onChange function for valid Date inputs, and will reset to the last valid Date input when the input loses focus.

To answer your question about the placement of the "X" button in the input pills, I changed it from the right side in the designs to be on the left side so it was smoother for the user to tab to the next input. Having the button on the right side meant the user needs to do an extra tab to get passed the close button, so I wanted to optimize for the flow where you are entering multiple filters and not worry about skipping over the button. Moving it to the left was just an easy way to get around the extra tab press.

Let me know if you have any more questions or feedback!

jfmaggie commented:
thanks for the fix! 👍 

got it!
